Use @fastify/fastify-postgres. CheckerNetwork/roadmap#220

Goddhi opened the pull request:
Added fastify and @fastify/postgres as dependencies
Updated stats/bin/spark-stats.js to use Fastify's PostgreSQL plugin
Improved connection error handling and graceful shutdown

EDIT @juliangruber:
For CheckerNetwork/roadmap#220

Goddhi committed:
fastify-postgres--implementation

pyropy commented on stats/lib/db.js:
```diff
@@ -0,0 +1,113 @@
+ * @param {string} options.evaluateConnectionString - Evaluate DB connection string
+ * @returns {Promise<void>}
+ */
+export async function initializeDb({ statsConnectionString, evaluateConnectionString }) {
```
Good job on extracting database logic to separate function 👏🏻

pyropy reviewed (changes requested):
Thank you for your submission @Goddhi 🚀 

Good start, but there's some changes that need to be addressed:

- Install `@fastify/postgres` package in correct workspace
- Revert or rethink on how the migrations are done
- Avoid creating new `Fastify` instances in order to register databases
- Figure out how to perform queries from fastify handlers

bajtos reviewed (changes requested):
+1 to what @pyropy wrote

Goddhi committed:
fastify/postgress-implementation

Goddhi commented:
@pyropy  I do appreciate your reviews, i have made the following changes as requested, kindly look into thanks.
Moved @fastify/postgres dependencies to correct workspace
make the  database connections to a single Fastify instance
removed the db.js which implements multiple Fastify instances. e.t.c.

pyropy commented on stats/lib/app.js:
```diff
@@ -12,18 +13,38 @@ import { addPlatformRoutes } from './platform-routes.js'
+    name: 'evaluate',
+  })
+
+  const pgPools = {
+    stats: app.pg.stats,
+    evaluate: app.pg.evaluate,
+    async end() {
+      await app.close()
+    }
+  }
```
```suggestion
```

I don't see this being used anywhere. If it's not used, let's removed.  [YAGNI](https://www.geeksforgeeks.org/what-is-yagni-principle-you-arent-gonna-need-it/)

juliangruber replied:
This is being passed to `addRoutes()` and `addPlatformRoutes()` below. However, instead of passing `pgPools`, the route handlers should access `app.pg.stats` etc. directly

pyropy commented on stats/bin/migrate.js:
```diff
@@ -1,9 +1,28 @@
-const pgPools = await getPgPools()
-await migrateStatsDB(pgPools.stats)
-await migrateEvaluateDB(pgPools.evaluate)
+const {
+  DATABASE_URL,
+  EVALUATE_DB_URL
+} = process.env
```
```suggestion
```
These constants seem not to be used anywhere. Let's delete them.

pyropy commented on stats/bin/migrate.js:
```diff
@@ -1,9 +1,28 @@
+  console.log('Running migrations for stats database...')
+  const pgPools = await getPgPools()
+  
+  // @ts-ignore - PgPoolStats actually does have a query method at runtime
```
Let's try to add correct types to migrate functions rather then using `@ts-ignore`. Using it should be avoided.

pyropy commented on stats/bin/migrate.js:
```diff
@@ -1,9 +1,28 @@
+  // @ts-ignore - PgPoolStats actually does have a query method at runtime
+  await migrateStatsDB(pgPools.stats)
+  
+  // @ts-ignore - Similarly for evaluate
```
Likewise, let's please not use `@ts-ignore`

pyropy commented on stats/bin/migrate.js:
```diff
@@ -1,9 +1,28 @@
+  
+  await pgPools.end()
+  
+  console.log('All migrations completed successfully')
```
Nice job on adding some logging 👏🏻

pyropy reviewed (changes requested):
Let's try to avoid having dead code around. Also, let's try to use new way of accessing databases in the API handlers.

pyropy commented:
@Goddhi I suggest you link the issue you're working on in the PR description (and not in title). In this case you should've linked this issue https://github.com/CheckerNetwork/roadmap/issues/220

juliangruber commented on stats/bin/migrate.js:
Why was this file changed? I don't see anything related to using fastify-postgres

juliangruber commented on stats/bin/spark-stats.js:
```diff
@@ -1,23 +1,26 @@
-const baseUrl = app.listen({ port: Number(PORT), host: HOST })
-console.log(baseUrl)
+await app.listen({ port: Number(PORT), host: HOST })
+console.log(`Server listening at ${HOST}:${PORT}`)
```
This change is unrelated, can you please revert it?

Goddhi committed:
FIX:integrated fastify/postgress implementation into routes

Goddhi commented:
@pyropy and @juliangruber have made commits that addresses the reviews.
I have updated route handlers to access database connections via request.server.pg which is a property  from the @fastify/postgres and as well simplified my code by removing unnecessary abstractions as suggested.

juliangruber commented on stats/lib/platform-routes.js:
```diff
@@ -14,39 +14,56 @@ import { filterPreHandlerHook, filterOnSendHook } from './request-helpers.js'
     app.get('/participants/top-earning', async (/** @type {RequestWithFilter} */ request, reply) => {
-      reply.send(await fetchTopEarningParticipants(pgPools.stats, request.filter))
-    })
+      const pgPools = adaptPgPools(request.server.pg);
+      reply.send(await fetchTopEarningParticipants(pgPools.stats, request.filter))    })
```
```suggestion
      reply.send(await fetchTopEarningParticipants(request.server.pg.stats, request.filter))    })
```

What do we need the `adaptPgPools` function for here?

juliangruber commented on stats/lib/routes.js:
```diff
@@ -23,57 +23,89 @@ import {
     app.addHook('onSend', filterOnSendHook)
 
     app.get('/deals/daily', async (/** @type {RequestWithFilter} */ request, reply) => {
+      const pgPools = adaptPgPools(request.server.pg);
```
what about instead changing the signature of `fetchDaily...` etc to accept `request.server.pg`? `fetchDailyDealStats(request.server.pg, request.filter)` etc

Goddhi replied:
alright, noted, i will do that now

pyropy replied:
+1 on changing the signature and dropping the `adaptPgPools` function.

pyropy reviewed (commented):
You're making great progress @Goddhi! 🚀

Goddhi committed:
FIX: modified fetcher functiond signatures to accept Fastify's pg object directly

Goddhi commented:
@juliangruber and @pyropy i have modified the fetcher functions signatures to accept Fastify's pg object directly and removed the adaptPgPools as suggested.

juliangruber commented on db/package.json:
```diff
@@ -13,12 +13,12 @@
     ]
   }
-}
+}
```
```suggestion
}

```

juliangruber commented on stats/bin/migrate.js:
```diff
@@ -6,4 +6,4 @@ import {
 const pgPools = await getPgPools()
 await migrateStatsDB(pgPools.stats)
-await migrateEvaluateDB(pgPools.evaluate)
+await migrateEvaluateDB(pgPools.evaluate)
```
```suggestion
await migrateEvaluateDB(pgPools.evaluate)

```

juliangruber commented on stats/lib/app.js:
```diff
@@ -2,28 +2,43 @@ import * as Sentry from '@sentry/node'
+  await app.register(fastifyPostgres, {
+    connectionString: EVALUATE_DB_URL,
+    name: 'evaluate',
+    
+    
```
```suggestion
```

juliangruber commented on db/package.json:
please resolve conflicts by merging in `main` and running `npm install`

juliangruber commented on stats/lib/platform-stats-fetchers.js:
```diff
@@ -1,16 +1,16 @@
 
 /**
- * @param {Queryable} pgPool
+ * @param {Object} pg 
```
```suggestion
 * @param {Queryable} pg 
```

I believe this still follows the `Queryable` interface. `Object` is too unspecific.

juliangruber commented on stats/lib/stats-fetchers.js:
```diff
@@ -1,12 +1,11 @@
-/** @typedef {import('@filecoin-station/spark-stats-db').PgPools} PgPools */
 /**
- * @param {PgPools} pgPools
+ * @param {Object} pg - Fastify pg object with database connections
```
Can you please find or create a more precise type than `Object`?

Goddhi replied:
I tried introducing a  type(non-generic) as you have suggested,  but it throws an error in the route handler becuase in Fastify with the @fastify/postgres plugin, request.server.pg is an object created by the plugin with a somewhat complex structure.  so can i go ahead with using the object type or  can you kindly recommend another way to implement this, thanks.

juliangruber replied:
This works for me:


```js
import fastifyPostgres from '@fastify/postgres'

/**
 * @param {fastifyPostgres.PostgresDb & Record<string, fastifyPostgres.PostgresDb>} pg - Fastify pg object with database connections
 * @param {import('./typings.js').DateRangeFilter & {nonZero?: 'true'}} filter
 */
```

That type `fastifyPostgres.PostgresDb & Record<string, fastifyPostgres.PostgresDb>` can be put in central place `stats/typings.d.ts` so it doesn't need to be repeated all over

Goddhi replied:
Alright, i have implemented it as you've suggested which worked and as well resolved merge conflicts issues.

Goddhi committed:
FIX: introduced FastifyPy type param

Goddhi committed:
Resolved merge conflicts in package-lock.json

juliangruber commented on stats/lib/platform-stats-fetchers.js:
```diff
@@ -1,16 +1,24 @@
+  @typedef {import('@filecoin-station/spark-stats-db').Queryable} Queryable 
+  @typedef {import('./typings.js').FastifyPg} FastifyPg 
+*/
+
+/**
+ * @param {FastifyPg} pg - Fastify pg object with database connections
+ */
```
```suggestion
```

juliangruber replied:
ping @Goddhi

juliangruber commented on stats/lib/stats-fetchers.js:
```diff
@@ -35,7 +41,7 @@ export const fetchRetrievalSuccessRate = async (pg, filter) => {
 
 /**
- * @param {Object} pg 
+ * @param {any} pg - Fastify pg object with database connections
```
```suggestion
 * @param {FastifyPg} pg - Fastify pg object with database connections
```

juliangruber commented on stats/lib/stats-fetchers.js:
```diff
@@ -399,7 +406,7 @@ export const fetchClientsRSRSummary = async (pg, filter) => {
 /**
  * Fetches the retrieval stats summary for a single client for given date range.
- * @param {Object} pg
+ * @param {{stats: Queryable, evaluate: Queryable}} pg
```
should this not also be `FastifyPg`?

juliangruber commented on stats/lib/typings.d.ts:
```diff
@@ -1,10 +1,16 @@
 import { FastifyRequest } from 'fastify'
+import { PostgresDb } from '@fastify/postgres'
+import { Queryable } from '@filecoin-station/spark-stats-db'
```
what is this type imported for?

juliangruber commented:
@Goddhi CI is failing, please take a look

Goddhi committed:
FIX: ci issue

Goddhi commented:
> @Goddhi CI is failing, please take a look

FIXED

juliangruber commented:
@Goddhi CI is still failing. You can run `npm test` locally to test this

Goddhi commented:
> @Goddhi CI is still failing. You can run `npm test` locally to test this

i have fixed other issues i found after running test, but the only issue left is this 
standard: Use JavaScript Standard Style (https://standardjs.com)
  /home/goddhi/spark-stats/stats/lib/platform-routes.js:2:3: 'fetchDailyStationCount' is defined but never used. (no-unused-vars)
  i also observed this handler function is also not used /stats/lib/platform-routes.js  of  the main branch, i dont know  if the CI will fail if i make commit now.

Goddhi committed:
FIX: added missing line of code

pyropy commented on stats/bin/spark-stats.js:
```diff
@@ -1,23 +1,26 @@
   SPARK_API_BASE_URL = 'https://api.filspark.com/',
-  REQUEST_LOGGING = 'true'
+  REQUEST_LOGGING = 'true',
+  DATABASE_URL,
+  EVALUATE_DB_URL
```
Let's add default values for both `env` variables:

```suggestion
  DATABASE_URL = 'postgres://localhost:5432/spark_stats',
  EVALUATE_DB_URL = 'postgres://localhost:5432/spark_evaluate'
```

pyropy commented on stats/lib/platform-routes.js:
```diff
@@ -14,39 +14,39 @@ import { filterPreHandlerHook, filterOnSendHook } from './request-helpers.js'
     })
     app.get('/stations/monthly', async (/** @type {RequestWithFilter} */ request, reply) => {
-      reply.send(await fetchMonthlyStationCount(pgPools.evaluate, request.filter))
+      reply.send(await fetchMonthlyStationCount(request.server.pg, request.filter))
```
```suggestion
      reply.send(await fetchMonthlyStationCount(request.server.pg.evaluate, request.filter))
```

We could also pass correct instance of the database instead of passing the `pg` object. In that case we don't need to change fetchers.

Goddhi replied:
This throws an error reason why went with only the pg object, besides the fetcher functions expects full pg object and uses pg.stats or pg.evaluate interally.

juliangruber replied:
What error does it throw?

juliangruber replied:
ping @Goddhi

pyropy commented on stats/lib/platform-stats-fetchers.js:
```diff
@@ -1,16 +1,24 @@
  */
-export const fetchDailyStationCount = async (pgPool, filter) => {
-  const { rows } = await pgPool.query(`
+export const fetchDailyStationCount = async (pg, filter) => {
```
In case we pass correct database instance we don't need to change this handler.

juliangruber replied:
ping @Goddhi

pyropy commented on stats/lib/routes.js:
```diff
@@ -23,58 +23,64 @@ import {
     })
     app.get('/retrieval-result-codes/daily', async (/** @type {RequestWithFilter} */ request, reply) => {
-      reply.send(await fetchDailyRetrievalResultCodes(pgPools, request.filter))
+      reply.send(await fetchDailyRetrievalResultCodes(request.server.pg, request.filter))
```
```suggestion
      reply.send(await fetchDailyRetrievalResultCodes(request.server.pg.stats, request.filter))
```

We could also pass correct instance of the database instead of passing the pg object. In that case we don't need to change fetchers.

juliangruber replied:
ping @Goddhi

juliangruber replied:
ping @Goddhi

pyropy reviewed (changes requested):
Great progress 👍🏻 

Let's clean the code up a bit; Let's try to pass database instance directly to the fetcher in which case we don't have to change fetcher signatures. 

Also I'd advise you to format your code using `standard` as that's what we use in majority of our repositories. You can do it simply by running `npx standard --fix`.

Goddhi committed:
FIX: CI test issues

Goddhi commented:
@pyropy and @juliangruber  i have made a commits regarding your suggestions, kindly help review it thanks.

juliangruber commented:
@Goddhi tests are still failing, please run `npm test` to verify locally. See the CI failures above

Goddhi commented:
> @Goddhi tests are still failing, please run `npm test` to verify locally. See the CI failures here

Yeah the reason the tests are failing is becuase i need to modify the test files to sync with the changes made in the fetcher and route files, this is what i am currently doing.

Goddhi commented:
https://github.com/CheckerNetwork/spark-stats/pull/341#discussion_r2009670371
@juliangruber this the errow that it throws request.server.pg.evaluate `Argument of type 'PostgresDb' is not assignable to parameter of type 'FastifyPg'.
  Type 'PostgresDb' is not assignable to type 'Record<string, PostgresDb>'.
    Property 'pool' is incompatible with index signature.
      Type 'Pool' is missing the following properties from type 'PostgresDb': pool, `

Goddhi committed:
FEAT: test implementation for fastify/postgres

Goddhi committed:
Merge remote-tracking branch 'upstream/main' into fastify-progres-implement
test

Goddhi committed:
FIX: test implementation for fastify/postgres

bajtos commented on stats/test/app.test.js:
```diff
@@ -37,18 +31,18 @@ describe('HTTP request handler', () => {
-    await pgPools.stats.query('DELETE FROM daily_scheduled_rewards')
-    await pgPools.stats.query('DELETE FROM daily_reward_transfers')
-    await pgPools.stats.query('DELETE FROM daily_retrieval_result_codes')
+    await app.pg.evaluate.query('DELETE FROM retrieval_stats')
```
Let's preserve `pgPools` structure please so that we don't have to change every source code line in tests that is working with the DB.

For example:

```js
before(async () => {
  app = await createApp({
    // ..
  })

  pgPools = app.pg
})

// no changes are needed in individual tests
```

Goddhi replied:
Alright, noted

bajtos commented on stats/lib/platform-routes.js:
```diff
@@ -48,6 +48,6 @@ export const addPlatformRoutes = (app) => {
   app.get('/participants/summary', async (request, reply) => {
     reply.header('cache-control', `public, max-age=${24 * 3600 /* one day */}`)
-    reply.send(await fetchParticipantsSummary(request.server.pg.evaluate))
+    reply.send(await fetchParticipantsSummary(request.server.pg))
```
Is this going to use server-wide connection or a per-request connection? Does it matter? Should we perhaps use something like `request.pg` instead? 

I am not familiar with the API provided by fastify-postgres. I just find it suspicious that request handlers are using something that looks like a single server-wide DB connection.

Goddhi replied:
Regarding your question @bajtos 
I ran tests using request.server.pg, and everything works as expected. @fastify/postgres provides a connection pool, so request.server.pg is not a single shared connection but rather an interface to the pool.

bajtos commented on stats/lib/stats-fetchers.js:
```diff
@@ -472,12 +472,12 @@ export const fetchAllocatorsRSRSummary = async (pgPools, filter) => {
 /**
  * Fetches the retrieval stats summary for a single allocator for given date range.
- * @param {PgPools} pgPools
+ * @param {FastifyPg} pg - Fastify pg object with database connections
```
Can we keep the name `pgPools` to keep the diff smaller?

```suggestion
 * @param {FastifyPg} pgPools - Fastify pg object with database connections
```

Same comment applies to all other similar places too.

Goddhi replied:
Alright, noted

juliangruber commented on stats/bin/migrate.js:
```diff
@@ -3,7 +3,6 @@ import {
   migrateEvaluateDB,
   migrateStatsDB
 } from '@filecoin-station/spark-stats-db'
-
```
please undo this change to keep the diff clean

Goddhi replied:
Alright, noted

Goddhi committed:
FIX: changed naming convention to pgPOols and CI test

Goddhi committed:
Merge remote-tracking branch 'upstream/main' into fastify-progres-implement
FIX: update to main

Goddhi committed:
FIX: linting issues

Goddhi commented:
@pyropy Please can you help check what could be causing the CI to fail?
i have fixed the line trailing issues, now the issue seems to be this "AssertionError [ERR_ASSERTION]: INFLUXDB_TOKEN required"
i guess the token needs to be updated from the action secret. Thanks
